### PR TITLE
dvclive: docs for log_param/log_params

### PR DESCRIPTION
**Pull request by dtrifiro**
documentation for `log_param`/`log_params` (https://github.com/iterative/dvclive/pull/292)

### CODE REVIEW

File: content/docs/dvclive/dvclive-with-dvc.md [hunk 1]
```diff
@@ -28,6 +28,10 @@ from dvclive import Live
 
 live = Live("training_metrics")
 
+live.log_param("num_epochs", NUM_EPOCHS)
```
Comment (dberenbaum) [inline content/docs/dvclive/dvclive-with-dvc.md:31]: I suggest not touching the `DVCLive with DVC` page until we complete some more steps to better integrate it into DVC. I don't think it makes much sense right now under stage params.

File: content/docs/dvclive/api-reference/live/index.md [hunk 2]
```diff
@@ -72,6 +75,8 @@ other metadata.
 ## Methods
 
 - `Live.log()`
+- `Live.log_param()`
+- `Live.log_params()`
```
Comment (dberenbaum) [inline content/docs/dvclive/api-reference/live/index.md:79]: I think you need to add these to `sidebar.json`.

File: content/docs/dvclive/api-reference/live/index.md [hunk 3]
```diff
@@ -38,6 +38,9 @@ other metadata.
 - `dir` - Location of the directory to store
   [outputs](/doc/dvclive/get-started#outputs).
 
+- `params_path` - `{Live.params_path}`. Location of the parameters logged via
+  [`log_param`](/doc/dvclive/api-reference/live/log_param#description)/[`log_params`](/doc/dvclive/api-reference/live/log_params#description).
+
 - `summary_path` - `{Live.dir}.json`. Location of the
```
Comment (dberenbaum) [inline content/docs/dvclive/api-reference/live/index.md:44]: cc @daavoo Should this be called a "metrics summary" now to distinguish from params? Should we update the attribute name?

File: content/docs/dvclive/dvclive-with-dvc.md (hunk 1 above)
Comment (dberenbaum) [reply content/docs/dvclive/dvclive-with-dvc.md:31]: Instead, I think this material should be added to `Get Started`.

File: content/docs/dvclive/api-reference/live/index.md [hunk 4]
```diff
@@ -38,6 +38,9 @@ other metadata.
 - `dir` - Location of the directory to store
   [outputs](/doc/dvclive/get-started#outputs).
 
+- `params_path` - `{Live.params_path}`. Location of the parameters logged via
```
Comment (dberenbaum) [inline content/docs/dvclive/api-reference/live/index.md:41]: Should it be `{Live.params_path}.yaml` to match `summary_path` (sorry this wasn't raised in the core PR)?

File: content/docs/dvclive/api-reference/live/log_params.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
+# Live.log_params()
+
+Logs the given group of parameters `params`, saving them as yaml.
```
Comment (dberenbaum) [inline content/docs/dvclive/api-reference/live/log_params.md:3]: Not sure it's worth pointing out the yaml format at the top, although making it clear somewhere in this doc makes sense.

File: content/docs/dvclive/api-reference/live/log_params.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    "lr": 1e-3,
+    "n_iter": 50
+}
+live.log_params(params)
+```
+
+## Description
+
+On each `live.log_params(params)` call, DVCLive will write keys/values pairs in
+the `params` dict to `{Live.params_file}`:
```
Comment (dberenbaum) [inline content/docs/dvclive/api-reference/live/log_params.md:28]:
```suggestion
the `params` dict to `{Live.params_path}`:
```

Also, if this becomes `{Live.params_path}.yaml`, I think that might be enough to indicate the output format.

File: content/docs/dvclive/api-reference/live/log_params.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+```py
+from dvclive import Live
+
+live = Live()
+
+params = {
+    "lr": 1e-3,
+    "n_iter": 50
+}
```
Comment (dberenbaum) [inline content/docs/dvclive/api-reference/live/log_params.md:21]: What do you think about including a nested param value here?

File: content/docs/dvclive/api-reference/live/log_params.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+## Parameters
+
+- `params` - Dictionary with name/value pairs of parameters to be logged.
+
+## Exceptions
+
+- `dvclive.error.InvalidParameterType` - thrown if any of the parameter values
+  in the `params` dict is not among supported types. Supported types include:
+  `{ParamLike=}`.
```
Comment (dberenbaum) [inline content/docs/dvclive/api-reference/live/log_params.md:51]: Seems like the ParamLike definition is missing here?

File: content/docs/dvclive/dvclive-with-dvc.md (hunk 1 above)
Comment (dtrifiro) [reply content/docs/dvclive/dvclive-with-dvc.md:31]: Agree

File: content/docs/dvclive/api-reference/live/log_params.md (hunk 7 above)
Comment (dtrifiro) [reply content/docs/dvclive/api-reference/live/log_params.md:21]: Yeah, that's fine

File: content/docs/dvclive/api-reference/live/index.md (hunk 3 above)
Comment (dtrifiro) [reply content/docs/dvclive/api-reference/live/index.md:44]: +1

Maybe we should also consider rethinking the summary/report names, they're somehow similar and it can be confusing

File: content/docs/dvclive/api-reference/live/log_params.md (hunk 6 above)
Comment (dtrifiro) [reply content/docs/dvclive/api-reference/live/log_params.md:28]: > Should it be {Live.params_path}.yaml to match summary_path (sorry this wasn't raised in the core PR)?
...
> Also, if this becomes {Live.params_path}.yaml, I think that might be enough to indicate the output format.

I'd like to have something like that, but I'd expect `dvclive.params_path` to be a full path (e.g. `dvclive/params.yaml`) Having something like `dvclive/params` would be pretty odd.

File: content/docs/dvclive/api-reference/live/log_params.md (hunk 7 above)
Comment (dtrifiro) [reply content/docs/dvclive/api-reference/live/log_params.md:21]: How about something like this? It showcases quite a few possibilities, although it's a bit large snippet

```python
params = {
    "num_classes": 10,
    "metrics": ["accuracy", "mae"],
    "training": {
        "epochs": 42,
        "batch_size": 128,
        "optimizer": {
            "name": "adam",
            "loss": "categorical_crossentropy",
            "learning_rate": 1e-3,
        },
    },
}
```

File: content/docs/dvclive/api-reference/live/index.md (hunk 4 above)
Comment (dberenbaum) [reply content/docs/dvclive/api-reference/live/index.md:41]: Sorry, got confused here. Can't it just be `params.yaml` since we got rid of the custom path?

Comment (dberenbaum) [reply content/docs/dvclive/api-reference/live/index.md:41]: In which case, does it even need to be a documented attribute/property?

Comment (dtrifiro) [reply content/docs/dvclive/api-reference/live/index.md:41]: I'd go with `{Live.dir}/params.yaml`

File: content/docs/dvclive/api-reference/live/log_params.md (hunk 8 above)
Comment (dtrifiro) [reply content/docs/dvclive/api-reference/live/log_params.md]: Added the definition, although I'm not sure it's worth it to have the exact definition in the docs. Maybe it will be enough to have something like:

> Supported types include basic python data types such as int, float, bool as well as maps and lists.

**Comment by dberenbaum** [comment]
@dtrifiro Let me know when it's ready for another review, thanks!

**Comment by dtrifiro** [comment]
@dberenbaum ready for review

File: content/docs/dvclive/api-reference/live/log_param.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+live = Live()
+
+live.log_param("lr", 1e-3)
+live.log_param("n_iter", 50)
+```
+
+<admon type="note">
+
+When using `Live(resume=True)`, calling `log_param()` after resuming a session
+will be a no-op.
```
Comment (daavoo) [inline content/docs/dvclive/api-reference/live/log_param.md:25]: Sorry for coming back to this at the docs stage, but what was the motivation for this?

Why having different params after `resume=True` is not considered valid?

cc @dberenbaum

Comment (dtrifiro) [reply content/docs/dvclive/api-reference/live/log_param.md:25]: I'd find it a bit odd to start one experiment with a set of parameters and resume it with another. In the context of experiment tracking, what set of parameters would be considered valid?

Comment (dtrifiro) [reply content/docs/dvclive/api-reference/live/log_param.md:25]: I agree having a no-op might be odd. Alternatives include
- overwriting `params.yaml` with the most recent one
- reading the existing `params.yaml` and comparing the parameters to the ones being logged and warning or failing when parameters do not match.
- ...

Comment (daavoo) [reply content/docs/dvclive/api-reference/live/log_param.md:25]: Note that `resume=True` was mainly introduced for `dvc checkpoints`, where one of the motivations is to resume from a checkpoint after modifying some parameters (https://dvc.org/doc/user-guide/experiment-management/checkpoints#starting-from-an-existing-checkpoint).

Outside `dvc checkpoints`, `resume=True` is just something I am not sure we should handle at all 😅, but we do. 

Your concern is valid but I guess we should let the users decide how to handle it. It is only a problem if you execute the training script without persisting the experiment.
If you `resume=True` from a persisted experiment, then the traceability is clear and it makes sense to allow overriding params.

Comment (daavoo) [reply content/docs/dvclive/api-reference/live/log_param.md:25]: > Alternatives include

@dtrifiro what if we change https://github.com/iterative/dvclive/blob/afa21a17c89c048a0e93c0dfbc7c57f8fbed5c83/src/dvclive/live.py#L230 to :

```python
if param_name in self._params and not self._resume:
```

Comment (dtrifiro) [reply content/docs/dvclive/api-reference/live/log_param.md:25]: This could be ok, although it might allow to log the same parameter twice when using resume.

Comment (dtrifiro) [reply content/docs/dvclive/api-reference/live/log_param.md:25]: This will only raise when logging the same parameter twice on step 0, although I feel like it's over-complicating this a bit
```python
            if param_name in self._params and not (
                self._resume and self.get_step()
            ):
```

Comment (dberenbaum) [reply content/docs/dvclive/api-reference/live/log_param.md:25]: > Note that `resume=True` was mainly introduced for `dvc checkpoints`, where one of the motivations is to resume from a checkpoint after modifying some parameters (https://dvc.org/doc/user-guide/experiment-management/checkpoints#starting-from-an-existing-checkpoint).

This is a good point, so if it's easy to accommodate, let's do it. I don't mind generally allowing users to override the dvclive values even if it could lead to dangerous/unexpected results. Users can mostly manage this themselves, and it's probably better to provide flexibility than enforce some strict ideas of correctness.

However, I don't feel the use case is that important, so let's not go crazy addressing this workflow. The idea of changing parameters mid-experiment has always seemed a little contrived to me and not something I have seen users asking for.

Comment (dberenbaum) [reply content/docs/dvclive/api-reference/live/log_param.md:25]: What would the behavior be if you remove https://github.com/iterative/dvclive/blob/afa21a17c89c048a0e93c0dfbc7c57f8fbed5c83/src/dvclive/live.py#L223-L227? Any parameters that are already logged would fail, but new parameters could still be logged (before or after resume)? I think that's fine if so. Again, I don't think we need to worry too much about either enforcing correctness or addressing every possible workflow for now.

Comment (dberenbaum) [reply content/docs/dvclive/api-reference/live/log_param.md:25]: Also fine: dropping `ParameterAlreadyLoggedError` altogether and leaving this entirely to users. I probably prefer that most. This kind of error is a little more important in metrics logging where users may forget to increment the step. 

Sorry for the lengthy responses here.

File: content/docs/dvclive/get-started.md [hunk 10]
```diff
@@ -22,6 +22,51 @@ live = Live()
 
 See [`Live()`](/doc/dvclive/api-reference/live) for details.
 
+### Log parameters
```
Comment (dberenbaum) [inline content/docs/dvclive/get-started.md:25]:
```suggestion
### (Optionally) Log parameters
```
Technically, it's all optional 😄 , but it might be worth making explicit that you don't have to log params.

File: content/docs/dvclive/get-started.md [hunk 11]
```diff
@@ -22,6 +22,51 @@ live = Live()
...
+live.log_param("training",
+    "epochs": 42,
+        "batch_size": 128,
+        "optimizer": {
+            "name": "adam",
+            "loss": "categorical_crossentropy",
+            "learning_rate": 1e-3,
+        },
+    }
+)
```
Comment (dberenbaum) [inline content/docs/dvclive/get-started.md:42]: All this detail is great for the API ref, but it's too much here. Let's leave it at the simple example to save space.

File: content/docs/dvclive/get-started.md [hunk 12]
```diff
@@ -22,6 +22,51 @@ live = Live()
...
+        "epochs": 42,
+        "batch_size": 128,
+        "optimizer": {
+            "name": "adam",
+            "loss": "categorical_crossentropy",
+            "learning_rate": 1e-3,
+        },
+    },
+}
+live.log_params(params)
```
Comment (dberenbaum) [inline content/docs/dvclive/get-started.md:62]: Again, we can simplify for get started. It doesn't need to be more than 3 lines at most.

File: content/docs/dvclive/get-started.md [hunk 13]
```diff
@@ -22,6 +22,51 @@ live = Live()
...
+```
+
+</tab>
+</toggle>
+
+See `Live.log_param()` and `Live.log_params()`.
```
Comment (dberenbaum) [inline content/docs/dvclive/get-started.md:68]: If we are doing the tabs, can we embed these links in the tabs like for `Log data` below?

File: content/docs/dvclive/get-started.md [hunk 14]
```diff
@@ -76,6 +121,10 @@ from dvclive import Live
 
 live = Live()
 
+model_params = {...}
+live.log_param(NUM_EPOCHS)
+live.log_params(model_params)
```
Comment (dberenbaum) [inline content/docs/dvclive/get-started.md:126]: Let's drop this. We don't need both param methods in this example.

File: content/docs/dvclive/api-reference/live/log_param.md (hunk 9 above)
Comment (dtrifiro) [reply content/docs/dvclive/api-reference/live/log_param.md:25]: >What would the behavior be if you remove [...] Any parameters that are already logged would fail, but new parameters could still be logged (before or after resume)? I think that's fine if so

The idea behind it is to be being able to have an idempotent training script, which you can run to resume training in case files already exist. If we drop that, we'll break this workflow (logging the same parameter twice will error out), unless we decide to drop `ParameterAlreadyLoggedError`, though I fear this might make for some annoying and hard to discover bugs when logging (by mistake) different values to the same key.

Comment (dberenbaum) [reply content/docs/dvclive/api-reference/live/log_param.md:25]: I think we are overthinking this and should start by dropping `ParameterAlreadyLoggedError`. I think users can handle checking whether they log different values to the same key and would not expect DVCLive to be responsible for checking this.

Comment (dtrifiro) [reply content/docs/dvclive/api-reference/live/log_param.md]: Got rid of it both here and in dvclive: https://github.com/iterative/dvclive/pull/316

File: content/docs/dvclive/api-reference/live/log_param.md [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+lr: 1e-3
+n_iter: 50
+```
+
+<admon type="tip">
+
+The parameters file `{Live.dir}/params.yaml` is usable by `dvc params`,
+`dvc metrics` and `dvc exp show`/`dvc exp diff`.
+
+</admon>
```
Comment (dberenbaum) [inline content/docs/dvclive/api-reference/live/log_param.md:38]: It's not that clear how to use it with any of these commands, or why we would suggest both `dvc params` and `dvc metrics`. What do you think about leaving this out until we add better support for tracking it in DVC?

File: content/docs/dvclive/api-reference/live/index.md [hunk 16]
```diff
@@ -71,10 +71,12 @@ other metadata.
...
 ## Methods
 
+- `Live.get_step()`
 - `Live.log()`
 - `Live.log_image()`
+- `Live.log_param()`
+- `Live.log_params()`
 - `Live.log_plot()`
 - `Live.make_report()`
-- `Live.get_step()`
```
Comment (jorgeorpinel) [inline content/docs/dvclive/api-reference/live/index.md:78]: Why was `get_step` moved up?

Comment (dberenbaum) [reply content/docs/dvclive/api-reference/live/index.md:78]: I think it was made alphabetical, but good catch that it doesn't match the sidebar. If you're fine with it, we can update the sidebar also. Probably easier for users to find things that way than trying to guess some order of importance.

Comment (jorgeorpinel) [reply content/docs/dvclive/api-reference/live/index.md:78]: Sure up to you. I was just pointing out a possible discrepancy I saw.
